Plot sequence logos

krish8484 opened the pull request:
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
This PR aims to plot the sequence logos using python's `logomaker` which may be used along with the heatmaps in the future.
Fixes #23 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

AngryMaciek commented on workflow/scripts/sequence_logos.py:
```diff
@@ -0,0 +1,97 @@
+input_files = options.input_files
+output_location = options.output_location
+
+sorted(input_files)
```
`sorted` in Python returns a new sorted object, not sorts in place, so you have to do: `input_files = sorted(input_files)`

krish8484 replied:
Right!

AngryMaciek commented on workflow/scripts/sequence_logos.py:
```diff
@@ -0,0 +1,97 @@
+###############################################################################
+#
+#   Script to plot sequence logos for motifss
```
motifss

krish8484 replied:
typo corrected

AngryMaciek commented on workflow/envs/plot_sequence_logos.yml:
```diff
@@ -0,0 +1,25 @@
+    - bioconda
+
+  dependencies:
+    - bash=4.4.18
```
Do you need `bash` here?

krish8484 replied:
No we dont, I have removed it.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -43,7 +43,17 @@ rule all:
+            ),
+            outdir = config["outdir"],
+            pwm_dir = gather_motifs_names(config["pwm_directory"])
+            )
```
4 spaces to the back for proper alignment :)

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -74,6 +84,57 @@ rule create_results_directory:
+            )
+
+    params:
+        DIR_sequence_logos = expand( 
```
`expand` is to be used under `input` of the following rule, not under `output` of the rule that is being parallelized. Think of it that there are 2 rules: former and latter. The latter executes the expansion and asks for many files. The former catches each of these request by providing output files in the form of wildcards.

krish8484 replied:
Currently, the output of the rule expects the `png` of all the pwm motifs, if not `expand`, how should we write it?

AngryMaciek replied:
This comment applies to both the expansions in the `params` and `output`.
Wildcards will flow from the subsequent rule (which in this case would be `all` rule) so, as I said, you do not use the keyword `expand` here. You may just specify the variables under output/params with pure wildcard variables: `variable = "{wildcard}"`

Please read on `expand` in snakemake.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -74,6 +84,57 @@ rule create_results_directory:
+    """
+    Create directories for the results
+    """
+    input:
```
Could you also link to the rule `create_output_directories` here? we need a dir for results before each of the other rules. We need a place to write to.

krish8484 replied:
Yes, you are right, linked!

krish8484 commented:
Are the sequence logos as expected?

AngryMaciek commented:
> Are the sequence logos as expected? They are in `output/sequence_logos`.

Dear Krish,

There is no directory called `output` anywhere I see, especially not uner the root of the repository... However, I have found some sample plots for HNRNPF undet unit test directory for this script so I will refer to these.
It's OK but I think of 3 changes:
* change `T` to `U` since it is a RNA sequence, not DNA. This is important.
* is the y-axis showing information content? I would exect that the library automatically plots information content but then the y-axis would range from 0 to 2. As it is not the case - could you please find out what is the scale and what does it mean? This is important.
* Could you remove the top and right-side lines from the "box"? Just for aesthetics, not that important.

krish8484 commented:
Hi Maciek,
The directory `output/sequence_logos` will be in `tests/integration` after you run the snakefile, though the `png` files in unit tests depict a similar output.

- I am not sure which `T` are you referring in the first point. If it is the one in `motif_HNRNPF_824.png`, then that `T` comes from the input file. See `motif_HNRNPF_824` in the same folder, the value of `T` for 1st position is `97.115`.
- The y axis represents the `probability` specified in the probability matrix in the input files (in unit tests folder). Hence the range 0 to 1.
- I could not find a function in the documentation which would remove the top and the right lines for us.

AngryMaciek commented:
* Yes, by convention we operate now on DNA sequence always, but for the plot please change the alphabet for `U`.
* Is there a flag in this library that would allow to plot information content instead of raw probabilities?
* `pyplot` has these options, just google: https://stackoverflow.com/questions/9750699/how-to-display-only-a-left-and-bottom-box-border-in-matplotlib I see that use are using `plt` to show so this solution should wotrk.

AngryMaciek commented:
If `logomaker` does not support plotting information content out-of-the-box, could you please calculate the contents youself? According to these specifications: https://en.wikipedia.org/wiki/Position_weight_matrix#Information_content
We do not need to deal with background probabilities, the base of `log` is `2`.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -344,13 +402,10 @@ rule plot_heatmap_of_MotEvo_results:
+        SVG_Heatmap_file = os.path.join(
+            "{outdir}",
+            "ProbabilityVsSequence.svg"
+            )
```
Wrong indentation, please look 4 lines above on hot wo match indent properly.

AngryMaciek commented on workflow/Snakefile:
```diff
@@ -74,6 +84,57 @@ rule create_results_directory:
+                os.path.join(
+                    "{outdir}",
+                    "outdir"
+                    ),
+        outdir = config["outdir"]
```
Wrong indentation, please stick to the formatting convention,

AngryMaciek commented:
As I was goung through the *Files changed* section here on GitHub and reviewing the code I noticed that under the dedicated directory for unit-testing this script you have also added the `png` sample output files to the repository (ex: `tests/unit/plot_sequence_logos/motif_HNRNPF_820.png`). Could you please remind me - do we need them in the repository? It just flew through my head that maybe it is not necessary to include them, since every unit test will generate them once again, right?

If that is correct - please remove the sample output files from unit testing of this script as well as previous unit tests: `combine_results`, `Plot-heatmap-for-motifs`,

krish8484 commented:
I have fixed the indentation in the Snakefile at the places you mentioned, removed the top and the right axes from the graph, changed the `T` to `U` in the input file in the unit tests, and changed the script to display `information` instead of `probability`.
Deleted the superfluous output files from unit tests and updated the md5sums of the output files.

AngryMaciek commented on tests/unit/plot_sequence_logos/motif_HNRNPF_824:
```diff
@@ -0,0 +1,9 @@
+//
+NA HNRNPF_824
+	A	C	G	U
```
Now you have an inconsistency: here you have a `U` but in other input files (eg: `tests/unit/plot_sequence_logos/motif_HNRNPF_823`) you have a column with `T`. Please keep it with `T` for now and just adjust the plotting with another alphabet.

AngryMaciek replied:
I am not sure if I made myself clear: all the input files should stay exaclty the way they are - please just encode in your plotting script one flag to use "U" instead of "T".

krish8484 replied:
Done, I have added a piece of code which will do this for us!

AngryMaciek commented:
Before I merge it please make sure that you have complied with all the items from the *Checklist:* in the initial post of the PR.

AngryMaciek commented:
@krish8484 
I have cleaned the code a little, indents (🙃), removed unnecessary expanions and **importantly** added two more sections to all your rules (which I forgot about earlier):
* `benchmark` - it allows *snakemake* to report the execution time per rule into a textfile
* `log` - to capture `stdout` and `stderr` streams of each script's execution into another two text files; very useful for debugging.
